program/src/processor.rs

```diff
@@ -820,6 +820,54 @@ fn process_switch_vote(
     Ok(())
 }
 
+/// Processes a
+/// [FinishVoting](enum.PaladinGovernanceInstruction.html)
+/// instruction.
+fn process_finish_voting(program_id: &Pubkey, accounts: &[AccountInfo]) -> ProgramResult {
+    let accounts_iter = &mut accounts.iter();
+
+    let proposal_info = next_account_info(accounts_iter)?;
+
+    check_proposal_exists(program_id, proposal_info)?;
+
+    let mut proposal_data = proposal_info.try_borrow_mut_data()?;
+    let proposal_state = bytemuck::try_from_bytes_mut::<Proposal>(&mut proposal_data)
+        .map_err(|_| ProgramError::InvalidAccountData)?;
+
+    // Ensure the proposal is in the voting stage.
+    if proposal_state.status != ProposalStatus::Voting {
+        return Err(PaladinGovernanceError::ProposalNotInVotingStage.into());
+    }
+
+    let clock = <Clock as Sysvar>::get()?;
+
+    match proposal_state.cooldown_timestamp {
```

Comment:
There's an edge case here that we should consider.

Since this (along with vote and switch-vote) is only checking that cooldown has elapsed, and not any of the thresholds, it's actually possible for a proposal to get into cooldown (acceptance threshold passed), people switch votes, it gets under the acceptance threshold, but *doesn't* go over the rejection threshold. With that, when the cooldown elapses, it'll pass despite not being over the acceptance threshold.

I can't remember if we already decided something about this, but how about adding a check that it still passes the acceptance threshold in all the places where the proposal can get moved to `Accepted`? Otherwise, it should be `Rejected`

Reply:
> Since this (along with vote and switch-vote) is only checking that cooldown has elapsed, and not any of the thresholds, it's actually possible for a proposal to get into cooldown (acceptance threshold passed), people switch votes, it gets under the acceptance threshold, but _doesn't_ go over the rejection threshold. With that, when the cooldown elapses, it'll pass despite not being over the acceptance threshold.

You're right, but I was under the impression this was the intended design. The cooldown begins whenever a vote or switched vote pushes stake for beyond the acceptance threshold. The only way to stop that cooldown, and convert an otherwise accepted proposal to rejected, is to cast or switch enough votes to bring it past the _rejection_ threshold.

We can change this, though, if you think it makes more sense. So in the case of a proposal getting enough support - say 50% - thus beginning the cooldown, and then 1% of votes switch so we now have 49% for (and say only 10% against), this is a rejected proposal when cooldown expires?

Reply:
Ah yeah, I think we did talk about this -- someone could maliciously vote to start the cooldown early and then immediately change their vote to kill a proposal otherwise. It might be worth asking Uri if this is OK or just documenting it somewhere

Reply:
As a follow-up to this conversation, it was decided by the Paladin team to implement the logic you're suggesting.

> someone could maliciously vote to start the cooldown early and then immediately change their vote to kill a proposal otherwise.

Paladin team is aware of this possibility and stated it's not a concern.